### PR TITLE
Adding an in memory object storage that can't be used in production.

### PR DESCRIPTION
**Pull request by eishay**
There is a InMemoryReportStoreImpl that's extending it. InMemoryReportStoreImpl will be used if the bucket name of the report storage is missing

### CODE REVIEW

File: shoebox/app/com/keepit/module/S3Module.scala [hunk 1]
```diff
@@ -52,10 +52,15 @@ class S3Module() extends ScalaModule with Logging {
...
-    val bucketName = S3Bucket(current.configuration.getString("amazon.s3.report.bucket").get)
-    new S3ReportStoreImpl(bucketName, amazonS3Client)
-  }
+  def reportStore(amazonS3Client: AmazonS3): ReportStore =
+    current.configuration.getString("amazon.s3.report.bucket") match {
+      case Some(name) =>
+        val bucketName = S3Bucket(name)
+        new S3ReportStoreImpl(bucketName, amazonS3Client)
+      case None =>
+        println(">> Using an InMemoryReportStoreImpl instead of a S3ReportStoreImpl")
```
Comment (2is10) [inline shoebox/app/com/keepit/module/S3Module.scala]: How about instead overriding this (original) binding in DevModule with one that uses `InMemoryReportStoreImpl` if the bucket name is unspecified? I guess it scares me to see `InMemoryReportStoreImpl` used in code that isn't explicitly for the dev environment.

File: shoebox/app/com/keepit/common/store/InMemoryObjectStore.scala [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import com.keepit.inject._
+import play.api.Play
+import play.api.Play.current
+import java.io.InputStream
+import java.io.ByteArrayInputStream
+import scala.collection.mutable.HashMap
+
+trait InMemoryObjectStore[A, B]  extends ObjectStore[A, B] with Logging {
+
+  if (Play.isProd) throw new Exception("Can't have in memory object store in production")
```
Comment (2is10) [inline shoebox/app/com/keepit/common/store/InMemoryObjectStore.scala:14]: wow, scary to only throw exceptions in production. how about sending a healthcheck alert instead?

**Comment by 2is10** [comment]
Thanks for doing this. Feel free to merge when you're happy with it. My comments were only brainstorms. Don't feel like you have to address them.

File: shoebox/conf/application.conf [hunk 3]
```diff
@@ -56,7 +56,8 @@ amazon.s3 = {
   article.bucket = "zarss-b-dev"
   social.bucket = "kakrafoon-b-dev"
   event.bucket = "arcturus-b-dev"
-  report.bucket = "betelgeuse-b-dev"
+  # If the report.bucket is missing then theInMemoryReportStoreImpl will be used if. Uncomment to use the s3 storage.
+  # report.bucket = "betelgeuse-b-dev"
   articleSearch.bucket = "article-search-results-b-dev"
```
Comment (2is10) [inline shoebox/conf/application.conf:61]: Would it make sense to do this for more or even all the buckets? I don't know what all of these are for. I didn't realize we consider s3 an important part of our development environment. If it is, then maybe this change isn't important after all.

Comment (eishay) [inline shoebox/conf/application.conf:61]: I don't think its super important to dev. We should be able to have most of dev running in isolated mode (say, not internet).
Yes, I think we should have that option to all buckets.

File: shoebox/app/com/keepit/common/store/InMemoryObjectStore.scala (hunk 2 above)
Comment (eishay) [inline shoebox/app/com/keepit/common/store/InMemoryObjectStore.scala:14]: Rather have a fast fail & rollback if such thing happen.

File: shoebox/app/com/keepit/module/S3Module.scala (hunk 1 above)
Comment (eishay) [inline shoebox/app/com/keepit/module/S3Module.scala]: good idea, i'll do it

File: shoebox/app/com/keepit/common/store/InMemoryObjectStore.scala (hunk 2 above)
Comment (andrewconner) [inline shoebox/app/com/keepit/common/store/InMemoryObjectStore.scala:14]: Does this make it so we can't run Play in prod mode locally?

Comment (eishay) [inline shoebox/app/com/keepit/common/store/InMemoryObjectStore.scala:14]: Its to make sure we won't accidentally use an in memory object store in prod.

File: shoebox/app/com/keepit/module/S3Module.scala (hunk 1 above)
Comment (eishay) [inline shoebox/app/com/keepit/module/S3Module.scala]: done

File: shoebox/conf/application.conf (hunk 3 above)
Comment (eishay) [inline shoebox/conf/application.conf:61]: done

File: shoebox/app/com/keepit/dev/S3DevModule.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+  def configure() {
+
+  }
+
+  @Singleton
+  @Provides
+  def articleSearchResultStore(amazonS3ClientProvider: Provider[AmazonS3]): ArticleSearchResultStore =
+    current.configuration.getString("amazon.s3.articleSearch.bucket") match {
+      case Some(name) => new S3Module().articleSearchResultStore(amazonS3ClientProvider.get)
```
Comment (2is10) [inline shoebox/app/com/keepit/dev/S3DevModule.scala:28]: Is there a way in Guice to delegate this binding to the overridden binding in the `Some(name)` case? (so we don't have to ensure that this one stays in sync with it)

**Comment by 2is10** [comment]
thanks again

File: shoebox/app/com/keepit/dev/S3DevModule.scala (hunk 4 above)
Comment (eishay) [inline shoebox/app/com/keepit/dev/S3DevModule.scala:28]: I didn't see anything about it in the docs

**Pull request merged**